File: Others/StarryCoding/Basic/Day12/P80 [NOIP2012 普及组] 摆花.py
from functools import cache

n, m = map(int, input().split())
A = list(map(int, input().split()))
MOD = int(1e6 + 7)

# @cache
# def dfs(i, j):
#     if i == n:
#         return 1 if j == 0 else 0
#     if j < 0:
#         return 0
#     res = 0
#     for k in range(A[i] + 1):
#         res = (res + dfs(i + 1, j - k)) % MOD
#     return res

# print(dfs(0, m))

dp = [[0] * (m + 1) for _ in range(n + 1)]
dp[0][0] = 1
for i in range(n):
    for j in range(m + 1):
        for k in range(A[i] + 1):
            if j - k < 0:
                break
            dp[i + 1][j] = (dp[i + 1][j] + dp[i][j - k]) % MOD
print(dp[n][m])

File: Others/StarryCoding/Basic/Day12/P80 [NOIP2012 普及组] 摆花.cpp
#include <bits/stdc++.h>
using namespace std;
const int MOD = 1e6 + 7;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, m;
    cin >> n >> m;
    vector<int> A(n);
    for (int i = 0; i < n; i++) cin >> A[i];

    vector<vector<int>> memo(n, vector<int>(m + 1, -1));
    auto dfs = [&](auto &&dfs, int i, int j) -> int {
        if (i == n) return (j == 0) ? 1 : 0;
        if (j < 0) return 0;
        if (memo[i][j] != -1) return memo[i][j];
        int res = 0;
        for (int k = 0; k <= A[i]; k++) {
            res = (res + dfs(dfs, i + 1, j - k)) % MOD;
        }
        return memo[i][j] = res;
    };
    cout << dfs(dfs, 0, m) << endl;
    return 0;
}